builtin/common/serialize.lua

```diff
@@ -1,205 +1,217 @@
+		else
+			reference[place] = succ[reference[place]]
 		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
+	end
+    -- [object] = reference string
+	local references = {}
+    -- Circular tables that must be filled using `table[key] = value` statements
+	local to_fill = {}
+	for object, count in pairs(count_objects(value)) do
+		local type_ = type(object)
+        -- Object must appear more than once. If it is a string, the reference has to be shorter than the string.
+		if count >= 2 and (type_ ~= "string" or #reference + 2 < #object) then
+			local ref = table_concat(reference)
+			write(ref)
+			write("=")
+            if type_ == "table" then
+                write("{}")
+            elseif type_ == "function" then
+                write(dump_func(object))
+            elseif type_ == "string" then
+                write(quote(object))
+            end
+			write(";")
+			references[object] = ref
+			if type_ == "table" then
+				to_fill[object] = ref
+			end
+			increment_reference(1)
 		end
-		-- First occurence, create and register reference
-		local val = dump_val(x)
-		local i = local_index
-		local_index = local_index + 1
-		var = "_["..i.."]"
-		local_defs[#local_defs + 1] = var.." = "..val
-		dumped[x] = var
-		return var
 	end
-
-	-- Second phase.  Dump the object; subparts occuring multiple times
-	-- are dumped in local variables which can be referenced multiple
-	-- times.  Care is taken to dump local vars in a sensible order.
-	function dump_val(x)
-		local  tp = type(x)
-		if     x  == nil        then return "nil"
-		elseif tp == "string"   then return string.format("%q", x)
-		elseif tp == "boolean"  then return x and "true" or "false"
-		elseif tp == "function" then
-			return string.format("loadstring(%q)", string.dump(x))
-		elseif tp == "number"   then
-			-- Serialize numbers reversibly with string.format
-			return string.format("%.17g", x)
-		elseif tp == "table" then
-			local vals = {}
-			local idx_dumped = {}
-			local np = nest_points[x]
-			for i, v in ipairs(x) do
-				if not np or not np[i] then
-					vals[#vals + 1] = dump_or_ref_val(v)
-				end
-				idx_dumped[i] = true
+	local function use_short_key(key)
+		return not references[key] and type(key) == "string" and string_match(key, "^[%a_][%a%d_]*$")
+	end
+	local function dump(value)
+		-- Primitive types
+		if value == nil then
+			return write("nil")
+		end
+		if value == true then
+			return write("true")
+		end
+		if value == false then
+			return write("false")
+		end
+		local type_ = type(value)
+		if type_ == "number" then
+			return write(string_format("%.17g", value))
+		end
+		-- Reference types: table, function and string
+		local ref = references[value]
+		if ref then
+			return write(ref)
+		end
+        if type_ == "string" then
+			return write(quote(value))
+		end
+        if type_ == "function" then
+			return write(dump_func(value))
+		end
+        if type_ == "table" then
+			local first = true
```

Comment:
It wouldn't really hurt to have a superficial ";" at the end of each table. And the code would be simpler and possibly faster.

Reply:
Hmm. I'm torn on that one. I don't like removing a feature... Guess I'll make it customizable in modlib, but here it is indeed unneeded.